docs: update Python 3.11 docs and more

robrap opened the pull request:
A variety of devstack doc updates, including:
- docs: manual upgrade for Python 3.11 
- docs: add note for Demo course is empty in studio
- fix: update readthedocs repo url
- fix: git openedx to edx repo docs

See https://github.com/edx/edx-arch-experiments/issues/578 for future work related to updating our docs.

----

I've completed each of the following or determined they are not applicable:

- [ ] Made a plan to communicate any major developer interface changes (or N/A)

rlucioni commented:
@clintonb @davec-edx @jmbowman @gsong FYI.

I still need to figure out what to do with the forwarded ports in our docker-compose file. I think our options are to remove the port forwarding altogether, or attempt to forward to different, unused ports. I don't know which is best. Thoughts?

rlucioni commented on loopback.sh:
```diff
@@ -0,0 +1,15 @@
+    sudo ifconfig lo0 alias 10.254.254.254
+
+    # This will check for a "docker.host" entry in your hosts file. It's created if it doesn't
+    # exist, so you can use URLs like "http://docker.host:8000" in your service configuration.
```
Hm, we might have to think harder about this one. If we want to use `docker.host` URLs in our configuration, I'll need to extend this script so that it writes to /etc/hosts on Linux, too.

rlucioni replied:
Addressed below.

davec-edx commented:
What do you mean? I'm not sure what the problem is.  Why can't the ports be
opened the way they are?

On Dec 9, 2016 11:36 PM, "Renzo Lucioni" <notifications@github.com> wrote:

> @clintonb <https://github.com/clintonb> @davec-edx
> <https://github.com/davec-edx> @jmbowman <https://github.com/jmbowman>
> @gsong <https://github.com/gsong> FYI.
>
> I still need to figure out what to do with the forwarded ports in our
> docker-compose file. I think our options are to remove the port forwarding
> altogether, or attempt to forward to different, unused ports. I don't know
> which is best. Thoughts?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/devstack/pull/20#issuecomment-266178531>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AWX4PLCrnYTbCiXwz1VOvLuaOlxEXYTsks5rGixJgaJpZM4LJjlP>
> .
>

gsong commented on README.md:
```diff
@@ -27,3 +27,22 @@ After the services have started, if you need shell access to one of the services
+
+Containers making requests to the LMS and Studio must communicate with ports exposed on the host system by the Vagrant VM.
+
+This is fine on Linux, but doesn't work out of the box on OS X. Attempting to access localhost on a container will result
```
Few points:

1. Why not use the [native Mac docker-engine](https://docs.docker.com/docker-for-mac/)?
2. A container's `localhost` in `net=bridge` mode will refer to the container itself, the host will have a different IP address. This is consistent across platforms.
3. Have you seen @nedbat's [write up about hooking up course-discovery container to Vagrant](https://openedx.atlassian.net/wiki/display/~ned/Running+Course+Discovery+alongside+Devstack)?

rlucioni replied:
1. I am; we should all be at this point. Although it doesn't rely on VirtualBox like the Docker Toolbox used to, Docker for Mac still uses a VM: HyperKit, a lightweight macOS virtualization solution built on top of Hypervisor.framework in macOS 10.10 Yosemite and higher ([reference](https://docs.docker.com/docker-for-mac/docker-toolbox/#/the-docker-for-mac-environment)).
2. I think this is what causes problems for us in accessing the LMS and Studio running in the Vagrant VM. The containers need to be able to access ports forwarded by the Vagrant VM. in order to do that they need to talk to the host itself, not themselves.
3. Yes. The approach here relies on the same loopback alias explored there.

rlucioni commented:
@davec-edx running our containers alongside the Vagrant VM causes port clashes. LMS and Studio forward several of the same ports (e.g., Elasticsearch on 9200).

clintonb commented on loopback.sh:
```diff
@@ -0,0 +1,19 @@
+
+    # This will check for a "docker.host" entry in your hosts file. It's created if it doesn't
+    # exist, so you can use URLs like "http://docker.host:8000" in your service configuration.
+    if ! grep -q "docker.host" /etc/hosts; then
```
Minor: Include a line in the README informing users how to _remove_ this if they want to (sadly) remove devstack.

rlucioni replied:
Added.

robrap committed:
docs: manual upgrade for Python 3.11

This is to allow a local devstack to continue to work
post this Python 3.11 update:
See https://github.com/openedx/edx-platform/pull/34982

robrap committed:
docs: add note for Demo course is empty in studio

Add update around ticket URL.

robrap committed:
fix: update readthedocs repo url

This probably won't fix readthedocs for us, but
may get us one step closer.

See https://github.com/edx/edx-arch-experiments/issues/578

robrap committed:
fix: git openedx to edx repo docs

Provide some fixes for working with the edx fork of devstack.

Also see https://github.com/edx/edx-arch-experiments/issues/578

robrap commented on docs/troubleshoot_general_tips.rst:
```diff
@@ -169,25 +169,38 @@ commands from within the repository::
+
+You can point to the 2U fork of devstack with the following command::
+
+    % git remote set-url origin ssh://git@github.com/edx/devstack
```
[inform] This is what I did locally, but my PR still got set up against openedx-unsupported at first, so there are other local changes I probably need to fix master, right?

timmc-edx replied:
That's a separate thing we'll need to fix. GitHub knows that we forked edx/devstack from openedx-unsupported/devstack (either before or after the move from openedx) and so it assumes we want to contribute PRs to the upstream. I don't know if there's a way to change that behavior other than asking GitHub to unlink the repo from that "network".

robrap replied:
That's a bummer.

robrap commented:
I added https://github.com/edx/edx-arch-experiments/issues/578#issuecomment-2187447978 to take note of other things I found. I also wonder if I should back out any non-RST changes from this PR so that provisioning is not required?

timmc-edx commented on docs/troubleshoot_general_tips.rst:
```diff
@@ -169,25 +169,38 @@ commands from within the repository::
+
+    % git remote set-url origin ssh://git@github.com/edx/devstack
+
+Note: You don't need to use ``ssh`` if you were using ``https` instead.
```
```suggestion
Note: You don't need to use ``ssh`` if you were using ``https`` instead.
```

timmc-edx commented:
I found the issue that `make docs` was failing on and added a suggestion.

But you could also remove the `make docs` from the end of the provisioning workflow (I don't think that was ever really needed there!) and add `docs/` to the `paths-ignore` of that workflow as well, if you wanted to do some clean there as well.

robrap committed:
feat: remove docs from provisioning tests

robrap committed:
fixup! fix quality issue
